lib/Dialect/TTIR/Transforms/HoistCPUOps.cpp

```diff
@@ -0,0 +1,234 @@
+#include "ttmlir/Dialect/TTIR/Transforms/Passes.h.inc"
+
+//===----------------------------------------------------------------------===//
+// Hoist CPU ops to standalone funcs pass
+//===----------------------------------------------------------------------===//
+static llvm::SmallVector<int64_t, 4> getTensorRanks(mlir::Operation *op) {
+  llvm::SmallVector<int64_t, 4> ranks;
+
+  // Iterate over operands (inputs)
+  for (auto operand : op->getOperands()) {
+    // Check if the operand is a tensor
+    if (auto tensorType = dyn_cast<mlir::RankedTensorType>(operand.getType())) {
+      // Add the rank of the tensor (number of dimensions)
+      ranks.push_back(tensorType.getRank());
+    }
+  }
+
+  // Iterate over results (outputs)
+  for (auto result : op->getResults()) {
+    // Check if the result is a tensor
+    if (auto tensorType = dyn_cast<mlir::RankedTensorType>(result.getType())) {
+      // Add the rank of the tensor (number of dimensions)
+      ranks.push_back(tensorType.getRank());
+    }
+  }
+
+  return ranks;
+}
+
+static std::string generateHoistedFuncName(mlir::Operation *op) {
+  std::string opName = op->getName().getStringRef().str();
+
+  // Start building the unique function name
+  std::string uniqueName = "hoisted_" + opName;
+
+  // Iterate over operands to extract tensor shapes and types
+  for (auto operand : op->getOperands()) {
+    mlir::ShapedType shapedType = dyn_cast<mlir::ShapedType>(operand.getType());
+    if (shapedType && shapedType.hasRank()) {
+      // Append the shape (dimensions) and the element type
+      std::string shapeStr = "_";
+      for (auto dim : shapedType.getShape()) {
+        shapeStr += std::to_string(dim) + "x";
+      }
+
+      // Append the element type (e.g., f32, i32) -- unforunately I don't think
+      // there's a better way to get string from mlir::Type
+      std::string elementTypeStr;
+      llvm::raw_string_ostream stream(elementTypeStr);
+      shapedType.getElementType().print(stream);
+
+      uniqueName += shapeStr + elementTypeStr;
+    }
+  }
+
+  // Add suffix to indicate it's a function
+  uniqueName += "_func";
+
+  return uniqueName;
+}
+
+static void hoistOperationToFunction(mlir::Operation *opToHoist,
+                                     mlir::ModuleOp sourceModule,
+                                     mlir::ModuleOp targetModule) {
+  const auto ranks = getTensorRanks(opToHoist);
+
+  const std::string functionName = generateHoistedFuncName(opToHoist);
+
+  auto localFunc = llvm::dyn_cast_or_null<func::FuncOp>(
+      sourceModule.lookupSymbol(functionName));
+
+  // if we have not already emitted an equivalent function call, perform this
+  // hoist
+  if (localFunc == nullptr) {
+
+    mlir::MLIRContext *context = sourceModule.getContext();
+
+    // Gather operand and result types
+    llvm::SmallVector<mlir::Type> operandTypes, resultTypes;
+    for (auto operand : opToHoist->getOperands())
+      operandTypes.push_back(operand.getType());
```

Comment:
Clang tidy is probably mad about lack of braces.

Reply:
Ha, I don't like that style either, I blame ChatGPT.  But I am suprised it doesn't get flagged by clang-format too...